### PR TITLE
Implement ObjectSafe and WF in the new solver

### PR DESCRIPTION
**Pull request by compiler-errors**
r? @lcnr

### CODE REVIEW

File: compiler/rustc_trait_selection/src/solve/mod.rs [hunk 1]
```diff
@@ -344,6 +348,35 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
...
+
+    fn compute_well_formed_goal(
+        &mut self,
+        goal: Goal<'tcx, ty::GenericArg<'tcx>>,
+    ) -> QueryResult<'tcx> {
+        self.infcx.probe(|_| {
+            match crate::traits::wf::obligations(
+                self.infcx,
+                goal.param_env,
+                CRATE_HIR_ID, // FIXME body id
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/solve/mod.rs:368]: This is kinda sketch but we only use this for the cause code, which we then.... throw away?

File: compiler/rustc_trait_selection/src/traits/wf.rs [hunk 2]
```diff
@@ -75,6 +75,30 @@ pub fn obligations<'tcx>(
     Some(result)
 }
 
+pub fn unnormalized_obligations<'tcx>(
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/wf.rs:78]: please add a comment that this should only be used in the new solver as it does not track span information.

File: compiler/rustc_trait_selection/src/traits/wf.rs [hunk 3]
```diff
@@ -75,6 +75,30 @@ pub fn obligations<'tcx>(
...
+    infcx: &InferCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    arg: GenericArg<'tcx>,
+) -> Option<Vec<traits::PredicateObligation<'tcx>>> {
+    if let ty::GenericArgKind::Lifetime(..) = arg.unpack() {
+        return Some(vec![]);
+    }
+
+    assert_eq!(infcx.shallow_resolve(arg), arg);
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/wf.rs:87]: could make this `debug_assert_eq!(infcx.resolve_vars_if_possible(arg), arg);` as it currently doesn't catch everything.

**Review by lcnr — COMMENTED** [review]
r=me after nits

**Pull request merged**